### PR TITLE
59 cosine hungarian

### PR DESCRIPTION
**Pull request by jspaaks**
Originally here https://github.com/matchms/matchms-backup/pull/246

So far we implemented a greedy version of the cosine score (which is often used because it is faster and gives the same results in 99,9% (that's a guess) of the actual cases.

The Hungarian algorithm would be the fully correct solution of the optimization problem, so it is nice to have it as well.

Here an example where the greedy scores would fail:
```python
from matchms import Spectrum
from matchms.similarity import CosineGreedy, CosineGreedyNumba, CosineHungarian

test_spectrum1 = Spectrum(mz=np.array([100.005, 100.016]),
                         intensities=np.array([1.0, 0.9]),
                         metadata={})

test_spectrum2 = Spectrum(mz=np.array([100.005, 100.01]),
                         intensities=np.array([0.9, 1.0]),
                         metadata={})

similarity_measure = CosineGreedyNumba(tolerance=0.01)
print("CosineGreedy:", similarity_measure(test_spectrum1, test_spectrum2))

similarity_measure = CosineGreedy(tolerance=0.01)
print("CosineGreedyNumba:", similarity_measure(test_spectrum1, test_spectrum2))

similarity_measure = CosineHungarian(tolerance=0.01)
print("CosineHungarian:", similarity_measure(test_spectrum1, test_spectrum2))
```
``CosineGreedy: (0.5524861878453039, 1)``
``CosineGreedyNumba: (0.5524861878453039, 1)``
``CosineHungarian: (0.994475138121547, 2)``

### CODE REVIEW

File: tests/test_cosine_hungarian.py [hunk 1]
```diff
@@ -0,0 +1,97 @@
+import numpy
```
Comment (cwmeijer) [inline tests/test_cosine_hungarian.py:1]: Could you some how clearify how you got to the 'expected' values in all the tests? Did you calculate that manually or something like that? I'm just thinking what one would think if any of these tests start to fail at some point and you are not there to explain what you meant.
If the calculation is simple and clear, you may want to include it in the test as opposed to these 'magic' ground truth values.

**Review by cwmeijer — CHANGES_REQUESTED** [review]
Nice addition. Good tests!

File: matchms/similarity/CosineHungarian.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+            assert max(spec1[:, 1]) <= 1, ("Input spectrum1 is not normalized. ",
+                                           "Apply 'normalize_intensities' filter first.")
+            assert max(spec2[:, 1]) <= 1, ("Input spectrum2 is not normalized. ",
+                                           "Apply 'normalize_intensities' filter first.")
+            return spec1, spec2
+
+        def get_matching_pairs():
+            """Get pairs of peaks that match within the given tolerance."""
+            matching_pairs = collect_peak_pairs(spec1, spec2, self.tolerance, shift=0.0)
+            matching_pairs = sorted(matching_pairs, key=lambda x: x[2], reverse=True)
```
[the remark this answers is not in the document]
Comment (florian-huber) [reply matchms/similarity/CosineHungarian.py]: Good point, thanks. Done.

File: matchms/similarity/CosineHungarian.py [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+
+class CosineHungarian:
+    """Calculate 'cosine similarity score' between two spectra (using Hungarian algorithm).
+
+    The cosine score aims at quantifying the similarity between two mass spectra.
+    The score is calculated by finding best possible matches between peaks
+    of two spectra. Two peaks are considered a potential match if their
+    m/z ratios lie within the given 'tolerance'.
+    The underlying peak assignment problem is here sovled using the Hungarian algorithm.
```
[the remark this answers is not in the document]
Comment (florian-huber) [reply matchms/similarity/CosineHungarian.py]: Thanks. Done.

File: matchms/similarity/CosineHungarian.py [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+        Args:
+        ----
+        spectrum1: SpectrumType
+            Input spectrum 1.
+        spectrum2: SpectrumType
+            Input spectrum 2.
+        """
+        def get_peaks_arrays():
+            """Get peaks mz and intensities as numpy array."""
+            spec1 = numpy.vstack((spectrum1.peaks.mz, spectrum1.peaks.intensities)).T
```
[the remark this answers is not in the document]
Comment (florian-huber) [reply matchms/similarity/CosineHungarian.py:42]: I kept it as it was right now. I will come back to it when working on #55 . Would that be OK for you?

File: tests/test_cosine_hungarian.py (hunk 1 above)
Comment (florian-huber) [reply tests/test_cosine_hungarian.py:1]: I changed the test cases to simpler ones where I now compare the actual scores with expected values that are derived within the same test (same as for you similar comment in #38 ).

File: matchms/similarity/CosineHungarian.py [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+
+        def get_matching_pairs():
+            """Get pairs of peaks that match within the given tolerance."""
+            matching_pairs = collect_peak_pairs(spec1, spec2, self.tolerance, shift=0.0)
+            matching_pairs = sorted(matching_pairs, key=lambda x: x[2], reverse=True)
+            return matching_pairs
+
+        def calc_score():
+            """Calculate cosine similarity score."""
+            used_matches = []
```
[the remark this answers is not in the document]
Comment (florian-huber) [reply matchms/similarity/CosineHungarian.py]: I tried to restructure the score calculation along your suggestions (I picked slightly different sub-functions though). Hope that makes it look better.

**Review by cwmeijer — APPROVED** [review]

**Comment by fdiblen** [comment]
Refs #6

**Comment by florian-huber** [comment]
Thanks @cwmeijer for the review!

**Pull request merged**